docs/ocis/adr/0011-global-url-format.md

```diff
@@ -0,0 +1,210 @@
+
+* Good, because bookmarks cannot break
+* Good, because URLs do not disclose unshared path segments
+* Bad, because URLs reveal no context to users
+
+### Path based URLs
+
+There is a customized ownCluod instance that uses path only based URLs:
+
+| URL | comment |
+|-|-|
+| `https://<host>/apps/files/?dir=/&` | root of the currently logged in user |
+| `https://demo.owncloud.com/apps/files/?dir=/&` | root of the currently logged in user |
+| `https://demo.owncloud.com/apps/files/?dir=/path/to/resource&` | sub folder `/path/to/resource` |
+
+* Good, because the URLs reveal the full path context to users
+* Good, because powerusers can navigate by updating the path in the url
+* Bad, because the bookmarks break when someone renames a folder in the path
+* Bad, because there is no id that can be used as a fallback lookup mechanism
+* Bad, because URLs might leak too much context (parent folders of shared files)
+
+### Space based URLs
+
+| URL | comment |
+|-|-|
+| `https://<host>/#/s/<space_id>(/<relative/path>)(?id=<resource_id>)` | the pattern, relative `path` and `resource_id` are optional |
+| `https://demo.owncloud.com/#/s/b78c2044-5b51-446f-82f6-907a664d089c:194b4a97-597c-4461-ab56-afd4f5a21607` | root of a storage space, might be the currently logged in users home |
+| `https://demo.owncloud.com/#/s/b78c2044-5b51-446f-82f6-907a664d089c:194b4a97-597c-4461-ab56-afd4f5a21607/relative/path/to/resource` | sub folder `/relative/path/to/resource` in the storage with id `b78c2044-5b51-446f-82f6-907a664d089c:194b4a97-597c-4461-ab56-afd4f5a21607`, works ***only*** if path still exists  |
+| `https://demo.owncloud.com/#/s/b78c2044-5b51-446f-82f6-907a664d089c:194b4a97-597c-4461-ab56-afd4f5a21607/relative/path/to/resource?id=ba4c1820-df12-11eb-8dcd-ff21f12c1264:beb78dd6-df12-11eb-a05c-a395505126f6` | sub folder `/relative/path/to/resource` in the storage with id `b78c2044-5b51-446f-82f6-907a664d089c:194b4a97-597c-4461-ab56-afd4f5a21607`, lookup can fall back to the `id` |
+
+{{< hint >}}
+* `/#` is used by the current vue router.
+* `/s` denotes that this is a space url.
+* `<space_id>` and `<resource_id>` both consist of `<storage_id>:<node_id>`, but the `space_id` can be replaced with a shorter id or an alias. See furthor down below.
+* `<relative/path>` takes precedence over the `<resource_id>`, both are optional
+{{< /hint >}}
+
+* Good, because the URLs reveal a relevant path context to users
+* Good, because everything after the `#` is not sent to the server, building the webdav request to list the folder is offloaded to the clients
+* Good, because powerusers can navigate by updating the path in the url
+* Bad, because the current ids are uuid based, leading to very long URLs where tha path component nearly vanishes between two very long strings
+* Bad, because the `#` in the URL is just a technical requirement
+* Bad, because ocis web requires a `/#/files/s` at the root of the route to distinguish the files app from other apps
```

Comment:
Hmmm... this is true. Would it make sense to assume the segment that comes after `/#/` is an app and if not present in a sort of "apps registry" treat it as a path for the files app? This is not ideal because we are introducing a twist in semantics to the URL. In my mind URLs should have only one way of doing things, and mixing behavior (like the one I just proposed) should be discouraged.

On a different note I don't see how namespacing apps is a bad argument against this types of URLs. Yes it clutters it, but it also does routing. URLs should not be 100% targeted to the user. Albeit nice to have 100% readable URLs there is only so much real state for user + business logic.

Reply:
having a `#` in the URL is a double edged sword. The `#` splits the _query parameters_ from the _anchor_, and the anchor is never transferred to the server. This may be desired or not. If it is not transferred the web ui has to take care of interpreting the anchor. If we don't use a `#` we can implement the redirect completely on the server side as a fallback. I have no strong opinion here, only that leaving it out given us more options at the cost of the anchor being sent to the server, potentially being looged on the server side ... which again might be a good or a bad thing.